Azure/azure-cli
# {Core} Honor expiresOn, instead of expiresIn

## Description

> Refine #15875, #15880

#12952 uses `expiresIn` to calculate the expiration date:

https://github.com/Azure/azure-cli/pull/12952/files#diff-7f489fcc6ae732563b52a550ccc3c32411ec118393eafd449ee9f5ab211ebf48R64

```py
return AccessToken(token, int(full_token['expiresIn'] + time.time()))
```

As tokens are cached, `expiresIn` saved in `~/.azure/accessTokens.json` will always be a fixed value, a little bit smaller or equal to `3600`. Using `full_token['expiresIn'] + time.time()` will result in an expiration time already past the actual expiration time (`expiresOn`).

```
Get token from server        3600s             Expiration
         |------------------------------------------|
               |------------------------------------------|
      Get token from cache                       Computed expiration
```

This way of computing expiration date is no different than "carving a mark on the boat to seek the sunk sword" (刻舟求剑).


## Changes

1. Honor `expiresOn`, instead of  `expiresIn`.
2. Add some token samples, to avoid future overlooked issue caused by different name of keys in the token entry (#14187) - User and service principal token entries have `expiresOn` in local time, while Managed Identity token entries has `expires_on` in epoch time.

## Testing Guide

Test on a local machine:

```sh
az login
az extension add -n account
account subscription show --id 0b1f6471-1bf0-4dda-aec3-cb9272f09590
```

Test on a VM with Managed Identity:

```sh
az login --identity
az extension add -n account
account subscription show --id 0b1f6471-1bf0-4dda-aec3-cb9272f09590
```

To test in Cloud Shell, see #13567.
